### ISSUE TITLE
PHP notice during feed processing when the line items setting is mapped to a single product field

### ISSUE THREAD

**Issue by richardW8k** [SproutApps/sprout-invoices-gravity-forms#3]
Selecting a product field with the field type set to "single product" results in two php notices during feed processing.  

> Notice: Undefined offset: 1 in /app/public/wp-content/plugins/sprout-invoices-gravity-forms/inc/SI_GF_Integration_Addon.php on line 85
> Notice: Undefined offset: 1 in /app/public/wp-content/plugins/sprout-invoices-gravity-forms/inc/SI_GF_Integration_Addon.php on line 86

The notices did not occur when a choice based product field, such as a drop down, was selected.

One way to resolve this would be to add the `field_type` property to the `products` setting and declare which field/input types should appear in the field mapping drop down.

An alternative approach would be to move that setting out of the field map and have it as a separate select type setting which only lists compatible fields and includes an additional choice allowing all the products/options from the entry order summary to be used.

`GFCommon::get_product_fields()` is used to populate the order summary on the entry detail page and when merge tags are processed, it could be used here to set the line items. Gravity Flow includes an example showing how it could be used to create a line items array here: https://github.com/gravityflow/gravityflow/blob/master/includes/steps/class-step-feed-slicedinvoices.php#L201

Using this method to get the products also means the line item quantity can be set to the quantity specified by the user when completing the form.

**Comment by dancameron** [comment]
I'll need to look into this more, thank you for pointing this out.

What direction would you recommend or benefit the most from?

**Comment by dancameron** [comment]
https://secure.helpscout.net/conversation/547735866/9910/?folderId=313618

**Comment by richardW8k** [comment]
I think the alternative approach would be the way to go.

I just noticed the file I linked to as an example has since changed so here is a new example showing how `GFCommon::get_product_fields()` can be used to populate a line items array: https://gist.github.com/richardW8k/155f11eddbd171caed0de045fd6f65c9

**Comment by dancameron** [comment]
Do you have an example of how to add a "separate select type" that can then be populated by these line item?  

FWIW I'm not sure I completely understand the approach you laid out now that I review it more in detail; I'm assuming it's create a new field type and then populate it. If that's the case I've actually tried this but could never get good enough documentation or examples to know how.

**Comment by richardW8k** [comment]
Take a look at 3586ec170ef6c76f5350fd7a1edae8d88c99ee32 and see what you think. 

What I have done is remove the existing `products` and `pd_line_items` settings and replaced them with a new `line_items` setting which is a select type field. This field is populated with the form product fields and an additional choice allowing them to use all the pricing fields. If SI Pre-defined Line Items is supported it will also contain an optgroup with the form checkbox fields.

Checkbox field population and feed processing have been updated to account for this change.

I also added an upgrade routine so existing feeds will be updated to use the new setting.

If this works for you I'll open a PR.
